Implement Metadata V2 (#341)
bodo-ai/PyDough: documentation/metadata.md

The core components of the JSON file structure are as follows:

- Each JSON file for PyDough metadata is a JSON array where each item is a JSON object representing a single knowledge graph.

> - Each knowledge graph object must have a `name` and `version` property, whose values are strings. The `name` property must be a valid Python identifier, and the `version` property can currently only be `V2` (indicating that the `V2` version of metadata should be used, which is the only current version of metadata).

Reviewer: ```suggestion
- Each knowledge graph object must have a `name` and `version` property, whose values are strings. The `name` property must be a valid Python identifier. The `version` must be set to `V2`—currently the only supported version.
```

> - Specification for the rest of the fields of the graph object if the version is `V2`:

Reviewer: ```suggestion
- Specification for the remaining fields when version is set to V2:
```

  - `relationships` (required): an array of the metadata for each relationship in the graph ([see here for more details](#relationships))
  - `additional definitions` (optional): an array of strings where each string is a sentence defining a concept/definition within the semantical context of the graph, such as the vocabulary for a specific kind of analysis and how to compute it in terms of the collections/relationships.
  - `verified pydough analysis` (optional): an array of JSON objects where each object represents an instance of a question & answer pair with PyDough using the data from the graph. The object has two fields that are both strings: `question` maps to the question being asked, and `code` is the PyDough code that solves the question using hte graph (can be multiline).
  - `extra semantic info` (optional): an object containing arbitrary additional semantic information about the entire graph (internal use only).
- Example of the structure of a metadata JSON file containing two graphs, both V2
```json
[
  {
    "name": "BankerGraph",
    "version": "V2",

>     "collections": [...],

Reviewer: should this include `accounts, clients, transactions` that were written in the previous example?

Author: Those would all be things inside this `"collections": [...]` list. They are no longer keys in a dictionary, now it would be to verbose to include in this small snippet.

  - Example C: `"unique properties": [["q1", "q2"]]` means every combination of values of properties `q1` and `q2` is unique.
  - Example D: `"unique properties": ["r1", ["r2", "r3", "r4"]]` means every value of property `r1` is unique, and every combination of values of properties `r2`, `r3` and `r4` is unique.
- `properties` (required): an array of objects representing scalar properties of each record from the collection ([see here for details](#properties)).

Example of the structure of the metadata for a simple table collection:

```json
{
    "name": "accounts",
    "type": "simple table",
    "table_path": "bank_schema.ACCOUNTS",
    "unique_properties": ["account_id"],
    "properties": {
        "account_id": {...},
        "client_id": {...},
        "account_type": {...},
        "account_balance": {...},
        "date_opened": {...}
    },
    "description": "Every account currently open in the bank's system",
    "extra semantic info": {...}
}
```

<!-- TOC --><a name="properties"></a>
## Properties

Every JSON object describing a scalar property of a PyDough collection ahs the following fields:

> - `name` (required): the name of the property, which must be a valid PyDough has and must be unique within the collection (and cannot conflict with the name of a relationship that the colleciton has)

Reviewer: ```suggestion
- `name` (required): the name of the property, which must be a valid PyDough identifier and must be unique within the collection (and cannot conflict with the name of a relationship that the collection has)
```

  - The code must be a single Python expression.
  - The code must return a PyDough expression, as opposed to a collection.
  - The code may call regular PyDough functions, but not window functions.
  - The code may not access any sub-collections of either collection.

Example of the structure of the metadata for a simple join property (connects a collection `Incidents` to a collection `WorkerShifts` by joining on whether `Incidents.incident_timestamp` is between `WorkerShifts.start_of_shift` and `WorkerShifts.end_of_shift`):

```json
{
    "name": "shifts_overlapping_with_incident",
    "type": "general join",
    "parent collection": "Incidents",
    "child collection": "WorkerShifts",
    "singular": false,
    "always matches": false,
    "condition": "MONOTONIC(other.start_of_shift, self.incident_timestamp, other.end_of_shift)",
}
```

<!-- TOC --><a name="relationship-type-cartesian-product"></a>
### Relationship Type: Cartesian Product

> A relationship with this type describes a subcollection of the current collection that is derived from performing an cross-join on two collections (e.g. SELECT ... FROM T1, T2). Relationships of this type are sub-collections of the collection, as opposed to scalar attributes. Relationships of this type have a type string of "cartesian product" and have the following additional key-value pairs in their metadata JSON object:

Reviewer: ```suggestion
A relationship with this type describes a subcollection of the current collection that is derived from performing a cross-join on two collections (e.g. SELECT ... FROM T1, T2). Relationships of this type are sub-collections of the collection, as opposed to scalar attributes. Relationships of this type have a type string of "cartesian product" and have the following additional key-value pairs in their metadata JSON object:
```

- `always matches` (optional): a boolean that is true if every record from the parent collection matches onto at least 1 record from the child, and false otherwise (default: `true`). This should always be true unless it is possible for the child collection to be empty.

Example of the structure of the metadata for a cartesian product property (connects every record of a collection `CalendarDates` to every record of collection `InventorySnapshots`):

```json
{
  "name": "snapshots",
  "type": "cartesian_product",
  "parent collection": "CalendarDates",
  "child collection": "InventorySnapshots",
  "always matches": true,
  "description": "Every single snapshot, accessed from every single calendar date",
  "synonyms": ["inventory logs", "storage records", "manifests"]
}
```

<!-- TOC --><a name="relationship-type-custom"></a>
### Relationship Type: Custom

> [!IMPORTANT]
> This type of metadata relationship has not yet been implemented.

<!-- TOC --><a name="relationship-type-reverse"></a>
### Relationship Type: Reverse

> A relationship with this type specifies that a relationship should be created by flipping the direction of an existing relationship that was already defined earlier in the `relationships` array. Relationships of this type are sub-collections of the collection, as opposed to scalar attributes. Relationships of this type have a type string of "cartesian product" and have the following additional key-value pairs in their metadata JSON object:

Reviewer: ```suggestion
A relationship with this type specifies that a relationship should be created by flipping the direction of an existing relationship that was already defined earlier in the `relationships` array. Relationships of this type are sub-collections of the collection, as opposed to scalar attributes. Relationships of this type have a type string of "reverse" and have the following additional key-value pairs in their metadata JSON object:
```

> - `original parent` (required): a string indicating the name of the collection who holds the relationship being reversed (this will be the child of hte new property)

Reviewer: ```suggestion
- `original parent` (required): a string indicating the name of the collection who holds the relationship being reversed (this will be the child of the new property)
```
